src/plugins/certauth/test/main.c

```diff
@@ -0,0 +1,220 @@
+ *   notice, this list of conditions and the following disclaimer in
+ *   the documentation and/or other materials provided with the
+ *   distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
+ * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+ * STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
+ * OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+#include <k5-int.h>
+#include "krb5/certauth_plugin.h"
+
+struct krb5_certauth_moddata_st {
+    int initialized;
+};
+
+/* Test module 1 returns OK with an indicator. */
+static krb5_error_code
+test1_authorize(krb5_context context, krb5_certauth_moddata moddata,
+                krb5_data cert, krb5_principal princ, void *opts,
+                void *db_entry, char ***authinds_out)
+{
+    krb5_error_code ret;
+    char **ais = NULL;
+
+    *authinds_out = NULL;
+    ais = k5calloc(2, sizeof(*ais), &ret);
+    if (ais == NULL)
+        return ret;
+
+    ais[0] = strdup("test1");
+    if (ais[0] == NULL) {
+        free(ais);
+        return ENOMEM;
+    }
+
+    *authinds_out = ais;
+    return 0;
+}
+
+static void
+test_free_ind(krb5_context context, krb5_certauth_moddata moddata,
+                                    char **authinds)
+{
+    size_t i;
+    if (authinds == NULL)
+        return;
+    for (i = 0; authinds[i] != NULL; i++) {
+        free(authinds[i]);
+    }
+    free(authinds);
+}
+
+/* A basic moddata test. */
+static krb5_error_code
+test2_init(krb5_context context, krb5_certauth_moddata *moddata_out)
+{
+    krb5_certauth_moddata mod;
+    mod = calloc(1, sizeof(*mod));
+    if (mod == NULL)
+        return ENOMEM;
+    mod->initialized = 1;
+    *moddata_out = mod;
+    return 0;
+}
+
+static void
+test2_fini(krb5_context context, krb5_certauth_moddata moddata)
+{
+   free(moddata);
+}
+
+static krb5_boolean
+has_cn(krb5_context context, krb5_data cert, char *name)
+{
+    krb5_boolean ret;
+    uint8_t cntag[6] = "\x06\x03\x55\x04\x03\x0c";
+    uint8_t *c, *cc;
+
+    /* Construct a DER search string of CN tag + length tag + name to
+     * check for princ as CN. */
+    c = malloc(sizeof(cntag) +  strlen(name) + 1);
+    if (c == NULL)
+        return FALSE;
+
+    memcpy(c, cntag, sizeof(cntag));
+    c[sizeof(cntag)] = strlen(name);
+    memcpy(c + sizeof(cntag) + 1, name, strlen(name));
+
+    cc = memmem(cert.data, cert.length, c, sizeof(cntag) + strlen(name) + 1);
```

Comment:
I think you'll want to use memchr() and memcmp() instead.  There will be a fiddly length check involved, but it shouldn't be a lot of code.